### PR TITLE
Implement `createWriteStream` for Linux

### PR DESCRIPTION
**Pull request by as-cii**
This pull request relies on [Polkit](https://wiki.archlinux.org/index.php/Polkit) to stream bytes into a file with administrator privileges. In particular it will use the `pkexec` command to invoke `dd` as an administrator, which in turn will display an authentication message similar to the following:

<p align="center">
Screen Shot 2019-05-28 at 13 39 57</p>

Users of `fs-admin` can customize this behavior by writing an appropriate `.policy` file in `/usr/share/polkit-1/actions`. Atom, for example, will include an `atom.policy` file that will display a custom message as well as retaining admin access to `dd` for a short period of time. See https://github.com/atom/atom/pull/19412 for more information.

### CODE REVIEW

File: index.js
Comment (shiftkey) [inline index.js]: Is there a chance that `/usr/bin/pkexec` or `/bin/pkexec` could be missing from the user's machine? If so, I'd love for this code to fail in a predictable way (or revert to the default behaviour)

**Review by shiftkey — COMMENTED** [review]
At first glance this all looks promising - thanks for the writeup!

> Users of `fs-admin` can customize this behavior by writing an appropriate `.policy` file in `/usr/share/polkit-1/actions`. 

We've not really needed a "packaging guide" for situations like this previously (because we relied on APIs included with the OS), but would you be interested in distilling this guidance into a document that we can add to the project? Linking off to the Atom PR in this document is also handy here, but I'd love something a bit more formal to explain what an app that distributes this library needs to be aware of...

File: index.js
Comment (as-cii) [reply index.js]: Good catch. ⚡️  It is indeed possible that some distros won't include Polkit. In Atom, we are working around that by requiring `polkit-1` to be a dependency of the Atom .deb/.rpm package.

That said, I was thinking we wouldn't need to do anything in particular here because `spawnSync` will fail with a non-zero status code and a `Failed to obtain credentials` exception will be reported to users of this API. 

The rationale for that is that I am assuming `createWriteStream` would only be used when it is  already known that the current user does not have access to a certain file. Is that assumption correct? 💭

**Comment by as-cii** [comment]
> We've not really needed a "packaging guide" for situations like this previously (because we relied on APIs included with the OS), but would you be interested in distilling this guidance into a document that we can add to the project? Linking off to the Atom PR in this document is also handy here, but I'd love something a bit more formal to explain what an app that distributes this library needs to be aware of...

Sounds good! Technically, including a policy would not be mandatory but it seems like a good nice-to-have. I'll write something up real quick! ✨

File: index.js
Comment (shiftkey) [reply index.js]: The reason why I've asked is that `spawnSync` and similar APIs don't exit cleanly with a non-zero exit code in some cases, and I've had to try-catch for `ENOENT` explicitly to avoid it bubbling up to the user.

Here's an example using an interactive Node session (Ubuntu 19.04):

```
$ node
Welcome to Node.js v12.3.1.
Type ".help" for more information.
> const { spawnSync } = require('child_process')
undefined
> spawnSync('/bin/missing')
{
  error: Error: spawnSync /bin/missing ENOENT
      at Object.spawnSync (internal/child_process.js:1041:20)
      at spawnSync (child_process.js:606:24)
      at repl:1:1
      at Script.runInThisContext (vm.js:123:20)
      at REPLServer.defaultEval (repl.js:384:29)
      at bound (domain.js:415:14)
      at REPLServer.runBound [as eval] (domain.js:428:12)
      at REPLServer.onLine (repl.js:700:10)
      at REPLServer.emit (events.js:205:15)
      at REPLServer.EventEmitter.emit (domain.js:471:20) {
    errno: 'ENOENT',
    code: 'ENOENT',
    syscall: 'spawnSync /bin/missing',
    path: '/bin/missing',
    spawnargs: []
  },
  status: null,
  signal: null,
  output: null,
  pid: 33370,
  stdout: null,
  stderr: null
}
> 
```

File: test/fs-admin.test.js
Comment (shiftkey) [inline test/fs-admin.test.js:73]: I won't hold up the PR for this change, but I've opened https://github.com/atom/fs-admin/issues/16 to think about how to better organize these tests now we have some improved Linux support.

File: index.js
Comment (as-cii) [reply index.js]: Oh interesting. Yeah I am relying on `status` being null in that circumstance, which will make the `status !== 0` conditional evaluate to true. Maybe there's a better way of structuring this?

Comment (shiftkey) [reply index.js]: Maybe this could be extracted out to it's own function, rather than trying to inline it?

```diff
diff --git a/index.js b/index.js
index 67b04d6..001d74c 100644
--- a/index.js
+++ b/index.js
@@ -151,6 +151,16 @@ switch (process.platform) {
     break
 
   case 'linux':
+
+    function spawnSyncWithoutError(command, args) {
+      const result = spawnSync(command, args)
+      if (result.error !== null) {
+        return false
+      }
+
+      return result.status !== 0
+    }
+
     Object.assign(fsAdmin, {
       clearAuthorizationCache () {
         spawnSync('/bin/pkcheck', ['--revoke-temp'])
@@ -158,7 +168,7 @@ switch (process.platform) {
 
       createWriteStream (filePath) {
         // Prompt for credentials synchronously to avoid creating multiple simultaneous prompts.
-        if (!fsAdmin.testMode && spawnSync('/usr/bin/pkexec', ['/bin/dd']).status !== 0) {
+        if (!fsAdmin.testMode && spawnSyncWithoutError('/usr/bin/pkexec', ['/bin/dd'])) {
           const result = new EventEmitter()
           result.write = result.end = function () {}
           process.nextTick(() => result.emit('error', new Error('Failed to obtain credentials')))
```

I'm not sure what other cases you would be looking to cover here, but that would at least let you keep the overall flow easy-to-read...

File: index.js
Comment (shiftkey) [inline index.js:175]: Might also need to handle `dd.on('error')` here and propagate that to `stream`, in case `/usr/bin/pkexec` generates an `ENOENT`, like above.

File: index.js
Comment (as-cii) [reply index.js]: Cool, I refactored it a bit in a7429bd. I ended up not extracting a function, but I think it reads better now. Curious to hear your thoughts about it! 😊

File: index.js
Comment (as-cii) [reply index.js:175]: I am not sure we need to worry about this, because the synchronous command invocation would have failed prevented `spawn` from having been called if `pkexec` did not exist. I guess there's a chance `pkexec` gets uninstalled after invoking `spawnSync` but that seems pretty unlikely. What do you think? 💭

**Review by as-cii — COMMENTED** [review]
Thanks for the promptly reviews, @shiftkey! 🙏 

I left a couple of comments, let me know what you think!

File: index.js
Comment (shiftkey) [reply index.js:175]: If you think they're either "both present" or "neither present" to the point that we can rely on that first check, then 👍 to not checking again here.

File: index.js
Comment (shiftkey) [reply index.js]: Works for me

**Review by shiftkey — APPROVED** [review]
LGTM @as-cii! I'm going to let this sit for the day to give @rafeca a chance to follow up too.

Planning to ship this as a 2.0.0 release (adding some proper Linux support feels like a good reason to bump the major version) later in the week if this is something you'd like to dogfood more in Atom.

**Comment by as-cii** [comment]
Thanks for the quick turnaround, @shiftkey! I think we may even bump the minor version since people shouldn't be relying on `createWriteStream` anyway on Linux. Happy to bump to 2.0 though if you think that'd be better! 👍 

Sounds good to wait for @rafeca to chime in before merging this 👍

**Comment by shiftkey** [comment]
> Happy to bump to 2.0 though if you think that'd be better! 👍

It's an additive API change rather than a true breaking change, so this isn't a strong opinion of mine.

**Review by rafeca — APPROVED** [review]
Looks good to me!

question: do we only need the `createWriteStream()` method for linux? (noticed that the other platform implementations have more methods e.g to create symlinks and stuff).

**Comment by shiftkey** [comment]
@rafeca 

> do we only need the `createWriteStream()` method for linux? (noticed that the other platform implementations have more methods e.g to create symlinks and stuff).

I'm not hugely fussed that this PR doesn't complete the full API surface. Given the underlying native module is an empty implementation on Linux I think there's still more work to be done here even if we reused the `darwin` JS code:

https://github.com/atom/fs-admin/blob/86dab87c0d7c7d8dd0b788a1047e41884dfe0ac4/src/fs-admin-linux.cc#L1-L19

I think the easiest option here is to open tracking issue for the API gaps so we're aware of what isn't currently supported. Is there any other APIs in here that you would need to support Atom's usage of `fs-admin` on Linux?

**Comment by as-cii** [comment]
> Is there any other APIs in here that you would need to support Atom's usage of `fs-admin` on Linux?

Not really, as we're planning to use this just for escalating privileges when saving a file. We may create issues to track other APIs but I am not sure we will need those in the short/medium term.

------

@shiftkey: please, feel free to merge/release this at will. Thanks again for the quick reviews! 🙇

**Comment by shiftkey** [comment]
@as-cii @rafeca this has been published to NPM as `v0.5.0`

**Pull request merged**
